frontends/common/resolveReferences/resolveReferences.cpp

```diff
@@ -28,6 +28,43 @@ static const std::vector<const IR::IDeclaration *> empty;
 
 ResolutionContext::ResolutionContext() { anyOrder = P4CContext::get().options().isv1(); }
 
+Util::Enumerator<const IR::IDeclaration *> *ResolutionContext::getDeclarations(
+    const IR::INamespace *ns) const {
+    auto nsIt = namespaceDecls.find(ns);
+    auto *declVec = nsIt != namespaceDecls.end() ? nsIt->second : memoizeDeclarations(ns);
+    return Util::Enumerator<const IR::IDeclaration *>::createEnumerator(*declVec);
+}
+
+std::vector<const IR::IDeclaration *> *ResolutionContext::memoizeDeclarations(
+    const IR::INamespace *ns) const {
+    auto decls = ns->getDeclarations();
+    if (auto nest = ns->to<IR::INestedNamespace>()) {
+        // boost bug -- trying to iterate with an adaptor over an unnamed temp crashes
```

Comment:
It is not a bug... it would just be a dangling reference. I know this comment existed before, but please update it to indicate it is to prevent dangling reference.

Reply:
I would also prefer to not use any boost in the first place here.

Reply:
What would be the preferred replacement for `boost::adaptors::reverse`?

Reply:
We could just add that adaptor ourselves, if we don't have it somewhere already. I actually happen to have it ready if you want to just take this code. Or I can contribute it in a separate PR


```cpp
/// A wrapper over a container that allows iterating it in reverse in a range for:
/// for (auto &x : reversed(someContainer)) { ... }
/// Note the name is deliberatly lower-case to look like a function as it is used like a function.
template<typename Container>
class reversed {
 public:
    static constexpr const bool is_const = std::is_const_v<std::remove_reference_t<Container>>;
    using iterator = std::conditional_t<is_const, typename Container::const_reverse_iterator,
                                                  typename Container::reverse_iterator>;
    using const_iterator = typename Container::const_reverse_iterator;
    using value_type = typename Container::value_type;
    using reference = std::conditional_t<is_const, typename Container::const_reference,
                                                   typename Container::reference>;
    using const_reference = typename Container::const_reference;

    /// Does NOT take owenership of the container, keeps reference only.
    explicit reversed(Container &container) : m_container(container) { }

    /// Prohibit copy and move, this is not intended to ever be stored.
    reversed(reversed &&) = delete;

    iterator begin() { return m_container.rbegin(); }
    iterator end() { return m_container.rend(); }
    const_iterator begin() const { return m_container.rbegin(); }
    const_iterator end() const { return m_container.rend(); }
    const_iterator cbegin() const { return m_container.crbegin(); }
    const_iterator cend() const { return m_container.crend(); }
 private:
    Container &m_container;
};
```

Reply:
I would prefer that you add it and replace this boost usage in a different PR, as this PR is about something else. But I can update the comment in this PR. Btw, `boost::adapter::reverse` is also used below in this file:
```
        // boost bug -- trying to iterate with an adaptor over an unnamed temp crashes
        auto temp = nested->getNestedNamespaces();
        for (auto nn : boost::adaptors::reverse(temp)) {
            auto rv = lookup(nn, name, type);
            if (!rv->empty()) return rv;
        }
```

Reply:
For now you could also just use a reverse_iterator `rbegin` `rend()` which are native.

Reply:
I can do that, but if @vlstill is going to replace this anyway with his `reversed` adapter in a different PR, then I'd rather just leave it as is.

Reply:
> I would also prefer to not use any boost in the first place here. We have been slowly trying to get rid of the boost dependencies in the compiler. #3898

This is a pre-exisinting code. Just moved into a helper. I would not add such change in the same commit for the sake of clarity. Boost could be removed elsewhere.